src/client/components/pages/Courses/tableFields.tsx

```diff
@@ -143,6 +148,60 @@ export const formatMeetings = (
     id: courseId,
   } = course;
   const [modalVisible, setModalVisible] = useState(false);
+  /**
+   * Keeps track of whether the user has altered fields in the form to determine
+   * whether to show a confirmation dialog on modal close
+   */
+  const [
+    isChanged,
+    setIsChanged,
+  ] = useState(false);
+
+  const confirmMessage = "You have unsaved changes. Click 'OK' to disregard changes, or 'Cancel' to continue editing.";
+
+  useEffect(() => {
+    /**
+     * Checks to see if there are any unsaved changes in the modal When the user
+     * refreshes the page. If there are unsaved changes, the browser displays a
+     * warning message to confirm the page reload. If the user selects cancel, the
+     * user can continue making changes in the modal.
+     */
+    const onBeforeUnload = (event: Event) => {
+      if (!isChanged) return;
+      event.preventDefault();
+      // It's unclear whether TS will account for browser incompatibility here,
+      // so we use all three methods of setting the confirmation message,
+      // as detailed here:
+      // https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event
+      const legacyEvent = (event as unknown as LegacyEvent);
+      legacyEvent.returnValue = confirmMessage;
+      return confirmMessage;
+    };
+    window.addEventListener('beforeunload', onBeforeUnload);
+    return () => {
+      window.removeEventListener('beforeunload', onBeforeUnload);
+    };
+  },
+  [isChanged]);
+
+  /**
+   * Called when the modal is closed. If there are any unsaved changes,
+   * a warning message appears, and the user must confirm discarding the unsaved
+   * changes in order to close the modal. If the user selects cancel, the user
+   * can continue making changes in the modal.
+   */
+  const onModalClose = () => {
+    if (isChanged) {
+      // eslint-disable-next-line no-alert
+      if (window.confirm(confirmMessage)) {
+        setIsChanged(false);
+        setModalVisible(false);
+      }
+    } else {
+      setModalVisible(false);
+    }
+  };
+
```

Comment:
@jonseitz My initial thought was to have this behavior in the `MeetingModal`, but I found that since the `modalVisible` state was in `tableFields` I started putting this logic in the `tableFields` file.

I started trying to move the `onModalClose` and `onBeforeUnload` logic into `MeetingModal` instead - one of the issues I found was that since I was conditionally setting `setModalVisible` to false depending on whether `isChanged` was true and whether the user clicked "OK" on the unsaved changes confirmation, I would need access to the `modalVisible` state within the `MeetingModal.` 

![image](https://user-images.githubusercontent.com/29589689/128216531-d6752ca2-876e-4ecb-af8b-d2aed34f5d77.png)

It seems like the `modalVisible` state should live in `tableFields,` since clicking an edit button sets the visibility to true and the `isVisible` property checks `modalVisible` to determine whether to show the modal or not. 

I passed down `setModalVisible` to the `Meeting Modal` so that we could control `setModalVisible` in `Meeting Modal`, but then I realized that `onModalClose` would be called `onClose` within `tableFields`:
![image](https://user-images.githubusercontent.com/29589689/128220078-8062ea47-797d-4952-9470-1f06b983e547.png)

I'm not really sure how to proceed here. Would it be helpful if I pushed up the code I have so far?

Reply:
I found a way to keep all of the behavior in the `MeetingModal` instead of implementing in the parent (5fa8eb3).